admin/compressor/class-boldgrid-backup-admin-compressor-system-zip.php

```diff
@@ -198,14 +206,131 @@ private function zip() {
 
 		$this->temp_folder->delete();
 
 		$this->core->logger->add( 'Finished closing the zip file.' );
 		$this->core->logger->add_memory();
 	}
 
+	/**
+	 * Get Filelist Chunks.
+	 *
+	 * @since SINCEVERSION
+	 *
+	 * @return array
+	 */
+	public function get_filelist_chunks() {
+		// Chunk size in bytes.
+		$max_chunk_size  = 26214400;
+		$size_of_chunk   = 0;
+		$filelist_chunks = array( array() );
+		$chunk_position  = 0;
+		foreach ( $this->filelist as $file ) {
+			// Adds file to this chunk.
+			$filelist_chunks[ $chunk_position ][] = $file[1];
+
+			// Adds the most recent file's size to chunk size.
+			$size_of_chunk = $size_of_chunk + (int) $file[2];
+
+			// If the chunk size is >= the max chunk size, then move to next chunk.
+			if ( $size_of_chunk >= $max_chunk_size ) {
+				$chunk_position++;
+				$size_of_chunk = 0;
+			}
+		}
+
+		return $filelist_chunks;
+	}
+
+	/**
+	 * Close.
+	 *
+	 * @since SINCEVERSION
+	 */
+	private function close() {
+		$chunks_closed   = 0;
+		$filelist_chunks = $this->get_filelist_chunks();
+		$total_chunks    = count( $filelist_chunks );
+
+		foreach ( $filelist_chunks as $filelist_chunk ) {
+			$chunk_start_time = microtime( true );
+			$add_file_string  = implode( ' ', $filelist_chunk );
+			$this->zip_proc( $filelist_chunk );
+			$chunks_closed++;
+			$percent_complete = round( $chunks_closed / $total_chunks, 2 );
+			$chunk_end_time   = microtime( true );
+			$close_duration   = $chunk_end_time - $chunk_start_time;
+			Boldgrid_Backup_Admin_In_Progress_Data::set_arg( 'percent_closed', $percent_complete );
+			$this->core->logger->add(
+				'Chunk closed in ' .
+				$close_duration .
+				' seconds. ' . $percent_complete * 100 .
+				'% complete closing'
+			);
+			$this->core->logger->add_memory();
+		}
+	}
+
+	/**
+	 * Get Compression Level.
+	 *
+	 * @since SINCEVERSION
+	 *
+	 * @return string
+	 */
+	private function get_compression_level() {
+		$default_compression_level = 6;
+		$compression_level         = $this->core->settings->get_setting( 'compression_level' );
+		return isset( $compression_level ) ? $compression_level : $default_compression_level;
+	}
+
+	/**
+	 * Close Zip using proc_open.
+	 *
+	 * @since SINCEVERSION
+	 *
+	 * @param array $filelist_chunk Array of Files to be added.
+	 */
+	private function zip_proc( $filelist_chunk ) {
+		$descriptorspec = array(
+			0 => array( 'pipe', 'r' ),  // stdin is a pipe that the child will read from.
+			1 => array( 'pipe', 'w' ),  // stdout is a pipe that the child will write to.
+			2 => array( 'file', '/tmp/error-output.txt', 'a' ), // stderr is a file to write to.
+		);
+
+		$cwd = ABSPATH;
+
+		$compression_level = $this->get_compression_level();
+
+		$process = proc_open( //phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.system_calls_proc_open
+			'zip -' . $compression_level . ' -g -q -@ ' . $this->filepath,
+			$descriptorspec,
+			$pipes,
+			$cwd
+		);
+
+		if ( is_resource( $process ) ) {
+			// $pipes now looks like this:
```

Comment:
Multiline comments should look like this - https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#5-2-multi-line-comments